### PR TITLE
Raise in JobQueueCluster._call when command returns with non zero exit code

### LINKED ISSUE #142: Should JobQueueCluster._call raise if the command has a non zero exit code?
I think there should be an exception if the submission command has a non zero exit code.

A slight side-effect of #132 is that you now get an exception later in the job id parsing which is not ideal.

This is on dask-jobqueue master using a queue that does not exist:
```py
from dask_jobqueue import SGECluster
from dask.distributed import Client
resource_spec = 'h_vmem=1000G,mem_req=16G'

cluster = SGECluster(queue='asdf',
                     cores=1,
                     memory='16GB',
                     resource_spec=resource_spec)
cluster.scale(2)
```

This is the output you get (on stderr):
```
Unable to run job: Job was rejected because job requests unknown queue "asdf".
Exiting.

tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x2aaac6552620>, 2)
Traceback (most recent call last):
  File "/sequoia/data1/lesteve/miniconda3/envs/dask-dev/lib/python3.6/site-packages/tornado/ioloop.py", line 760, in _run_callback
    ret = callback()
  File "/sequoia/data1/lesteve/miniconda3/envs/dask-dev/lib/python3.6/site-packages/tornado/stack_context.py", line 276, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/lesteve/dev/dask-jobqueue/dask_jobqueue/core.py", line 375, in scale_up
    self.start_workers(n - active_and_pending)
  File "/home/lesteve/dev/dask-jobqueue/dask_jobqueue/core.py", line 302, in start_workers
    job = self._job_id_from_submit_output(out.decode())
  File "/home/lesteve/dev/dask-jobqueue/dask_jobqueue/core.py", line 420, in _job_id_from_submit_output
    raise ValueError(msg)
ValueError: Could not parse job id from submission command output.
Job id regexp is '(?P<job_id>\\d+)'
Submission command output is:
```

So basically you get some warning that the submission output had some stderr (with `logger.error`) but then a traceback from the job_id parsing.

I think we should just have a traceback from `JobQueueCluster.__call__` saying that the submission command exited with a non zero exit code, and put the stdout and stderr to help the user figure out what happened.

Any comments let me know, otherwise I'll implement something like this.

### PR DESCRIPTION
**Pull request by lesteve**
Fix #142.

Here is the same snippet as in #142:
```py
from dask_jobqueue import SGECluster
from dask.distributed import Client
resource_spec = 'h_vmem=1000G,mem_req=16G'

cluster = SGECluster(queue='asdf',
                     cores=1,
                     memory='16GB',
                     resource_spec=resource_spec)
cluster.scale(2)
```

Error on **master**
```
Unable to run job: Job was rejected because job requests unknown queue "asdf".
Exiting.

tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x2aaac6552620>, 2)
Traceback (most recent call last):
  File "/sequoia/data1/lesteve/miniconda3/envs/dask-dev/lib/python3.6/site-packages/tornado/ioloop.py", line 760, in _run_callback
    ret = callback()
  File "/sequoia/data1/lesteve/miniconda3/envs/dask-dev/lib/python3.6/site-packages/tornado/stack_context.py", line 276, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/lesteve/dev/dask-jobqueue/dask_jobqueue/core.py", line 375, in scale_up
    self.start_workers(n - active_and_pending)
  File "/home/lesteve/dev/dask-jobqueue/dask_jobqueue/core.py", line 302, in start_workers
    job = self._job_id_from_submit_output(out.decode())
  File "/home/lesteve/dev/dask-jobqueue/dask_jobqueue/core.py", line 420, in _job_id_from_submit_output
    raise ValueError(msg)
ValueError: Could not parse job id from submission command output.
Job id regexp is '(?P<job_id>\\d+)'
Submission command output is:

```

Error with **this PR**
```
tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x2aaaeb9e3598>, 2)
Traceback (most recent call last):
  File "/sequoia/data1/lesteve/miniconda3/envs/dask-dev/lib/python3.6/site-packages/tornado/ioloop.py", line 760, in _run_callback
    ret = callback()
  File "/sequoia/data1/lesteve/miniconda3/envs/dask-dev/lib/python3.6/site-packages/tornado/stack_context.py", line 276, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/lesteve/dev/dask-jobqueue/dask_jobqueue/core.py", line 379, in scale_up
    self.start_workers(n - active_and_pending)
  File "/home/lesteve/dev/dask-jobqueue/dask_jobqueue/core.py", line 301, in start_workers
    out = self._submit_job(fn)
  File "/home/lesteve/dev/dask-jobqueue/dask_jobqueue/core.py", line 293, in _submit_job
    return self._call(shlex.split(self.submit_command) + [script_filename])
  File "/home/lesteve/dev/dask-jobqueue/dask_jobqueue/core.py", line 350, in _call
    cmd_str, out, err))
RuntimeError: Command exited with non-zero exit code.
Exit code: 1
Command:
qsub -terse /tmp/tmpj48t7zkx.sh
stdout:

stderr:
Unable to run job: Job was rejected because job requests unknown queue "asdf".
Exiting.
```

A few remarks:
* Because I was touching this code, I took this opportunity of removing `JobQueueCluster._calls` which calls multiple commands since we are only calling one command at a time. Maybe @mrocklin remembers why there was `_calls/_call` (maybe this is related to the comment "and an opportunity to go asynchronous in the future." that I don't fully understand if I am honest).
* I am using `.decode()` for stdout and stderr to have better error messages. If anyone sees a reason why that could be problematic, let me know. Note: on Python 3, `subprocess.Popen` + `communicate` returns bytes and not str.

### CODE REVIEW

**Comment by mrocklin** [comment]
In general I really like the improved error messages.

> I took this opportunity of removing JobQueueCluster._calls which calls multiple commands since we are only calling one command at a time. Maybe @mrocklin remembers why there was _calls/_call (maybe this is related to the comment "and an opportunity to go asynchronous in the future." that I don't fully understand if I am honest)

No objection from me

File: dask_jobqueue/core.py [hunk 1]
```diff
@@ -308,43 +310,48 @@ def scheduler(self):
...
+        This centralizes calls out to the command line, providing consistent
+        outputs, logging, and an opportunity to go asynchronous in the future.
 
         Parameters
         ----------
-        cmd: List(List(str))
-            A list of commands, each of which is a list of strings to hand to subprocess.communicate
+        cmd: List(str))
+            A command, each of which is a list of strings to hand to
+            subprocess.communicate
```
Comment (guillaumeeb) [inline dask_jobqueue/core.py]: `subprocess.Popen` ?

File: dask_jobqueue/core.py [hunk 2]
```diff
@@ -308,43 +310,48 @@ def scheduler(self):
         """ The scheduler of this cluster """
         return self.local_cluster.scheduler
 
-    def _calls(self, cmds, **kwargs):
-        """ Call a command using subprocess.communicate
+    def _call(self, cmd):
```
Comment (guillaumeeb) [inline dask_jobqueue/core.py]: You should keep **kwargs attribute for LSF specificities.

File: dask_jobqueue/core.py [hunk 3]
```diff
@@ -308,43 +310,48 @@ def scheduler(self):
...
-
-    def _call(self, cmd, **kwargs):
-        """ Singular version of _calls """
-        return self._calls([cmd], **kwargs)[0]
+        cmd_str = ' '.join(cmd)
+        logger.debug("Executing the following command to command line\n{}".format(cmd_str))
+
+        proc = subprocess.Popen(cmd,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
```
Comment (guillaumeeb) [inline dask_jobqueue/core.py]: Add **kwargs here, see previous comment.

**Comment by guillaumeeb** [comment]
@lesteve, looks like you've been busy with other things lately. Do you plan on getting back to dask-jobqueue (this would be quite welcomed :grinning: )? Or should we take over your ongoing work?

**Comment by lesteve** [comment]
Indeed I have been busy with other things lately, sorry about that ...

Looks like there are non-trivial merge conflicts in this PR, I'll try to revive it. This will very likely not happen this week but I'll try my best for next week.

**Comment by guillaumeeb** [comment]
> Looks like there are non-trivial merge conflicts in this PR, I'll try to revive it. This will very likely not happen this week but I'll try my best for next week.

So I've done the merge and corrected my remark, one final look from you @lesteve would be good.

**Comment by guillaumeeb** [comment]
Thanks @lesteve.

**Pull request merged**